### PR TITLE
Final Review of IntentMaxHistoryTrackerFeaturizer

### PR DESCRIPTION
**Pull request by kedz**
- Misc. doc string edits in IntentMaxHistoryTrackerFeaturizer
  - https://github.com/RasaHQ/rasa/pull/8913/files#r661351580
  - https://github.com/RasaHQ/rasa/pull/8913/files#r662279690
  - https://github.com/RasaHQ/rasa/pull/8913/files#r661508961
  - https://github.com/RasaHQ/rasa/pull/8913/files#r662298463
  - https://github.com/RasaHQ/rasa/pull/8913/files#r662313726
  - https://github.com/RasaHQ/rasa/pull/8913/files#r661360084
  - https://github.com/RasaHQ/rasa/pull/8913/files#r662422407
- ~~Split IntentMaxHistoryTrackerFeaturizer into a separate module. Split the corresponding test file as well.~~ 
  - https://github.com/RasaHQ/rasa/pull/8913/files#r661356109
- ~~Updated UnexpecTEDIntentPolicy and its tests to import IntentMaxHistoryTrackerFeaturizer correctly after module split.~~

update: no longer splitting the `tracker_featurizers` module code.

### CODE REVIEW

File: rasa/core/featurizers/intent_max_history_tracker_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+            for tracker_intents in trackers_as_intents
+        ]
+
+        return np.array(cls._pad_label_ids(label_ids))
+
+    @staticmethod
+    def _pad_label_ids(label_ids: List[List[int]]) -> List[List[int]]:
+        """Pads label ids so that all are of the same length.
+
+        The pad value `LABEL_PAD_ID` is set in `rasa.utils.tensorflow.constants`.
```
Comment (dakshvar22) [inline rasa/core/featurizers/intent_max_history_tracker_featurizer.py:62]: I think this statement is not needed in the docstring as it's evident from the import statement where it's defined and users can navigate. Moreover, if it gets refactored in the future to be placed somewhere else then I am sure we'll forget to modify this docstring 🙂

File: rasa/core/featurizers/tracker_featurizers.py [hunk 2]
```diff
@@ -524,15 +522,6 @@ def load(path: Text) -> Optional["TrackerFeaturizer"]:
 
     @staticmethod
     def _remove_action_unlikely_intent_from_states(states: List[State]) -> List[State]:
```
Comment (dakshvar22) [inline rasa/core/featurizers/tracker_featurizers.py:524]: I know it wasn't mentioned in the original PR review but I think we can remove the docstring for `_remove_action_unlikely_intent_from_events` as well.

File: rasa/core/featurizers/intent_max_history_tracker_featurizer.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+                label = [event.intent_name or event.text]
+                entities = [{}]
+
+                yield sliced_states, label, entities
+
+    @staticmethod
+    def _cleanup_last_user_state_with_action_listen(
+        trackers_as_states: List[List[State]],
+    ) -> List[List[State]]:
+        """Removes the last state in a tracker if the previous action is action_listen.
```
Comment (dakshvar22) [inline rasa/core/featurizers/intent_max_history_tracker_featurizer.py:223]:
```suggestion
        """Removes the last state in a tracker if the previous action is `action_listen`.
```

File: rasa/core/featurizers/intent_max_history_tracker_featurizer.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+        States with the previous action equal to action_listen correspond to states with
+        a new user intent. This information is what UnexpecTEDIntentPolicy is trying to
+        predict so it needs to be removed before obtaining a prediction.
```
Comment (dakshvar22) [inline rasa/core/featurizers/intent_max_history_tracker_featurizer.py:227]: Sam has an eagle eye for these things, so pointing them out here before he sees it in the final PR 😅 
```suggestion
        States with the previous action equal to `action_listen` correspond to states with
        a new user intent. This information is what `UnexpecTEDIntentPolicy` is trying to
        predict so it needs to be removed before obtaining a prediction.
```

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Just minor things to fix before we can merge.

File: rasa/core/featurizers/intent_max_history_tracker_featurizer.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
+from typing import List, Text, Tuple, Dict, Any, Iterator, Optional
+from collections import defaultdict
+from tqdm import tqdm
+import numpy as np
+import logging
+
+from rasa.core.featurizers.tracker_featurizers import (
+    MaxHistoryTrackerFeaturizer,
+    is_prev_action_listen_in_state,
```
Comment (wochinge) [inline rasa/core/featurizers/intent_max_history_tracker_featurizer.py:9]: can we please not import the function directly?

**Review by wochinge — COMMENTED** [review]
I think this actually breaks [this logic](https://github.com/RasaHQ/rasa/blob/f6125237c4b3d1a7d1489fffad2e4bf521e98e93/rasa/core/policies/ensemble.py#L460), doesn't it? 🙈

File: rasa/core/featurizers/tracker_featurizers.py [hunk 6]
```diff
@@ -177,6 +173,10 @@ def _entity_data(event: UserUttered) -> Dict[Text, Any]:
     def _remove_user_text_if_intent(trackers_as_states: List[List[State]]) -> None:
         """Deletes user text from state dictionaries if intent is present.
 
+        Only featurizing either the intent or user text is currently supported. When
+        both are present in a state, the user text is removed so that only the intent
+        is featurized.
+
```
Comment (kedz) [inline rasa/core/featurizers/tracker_featurizers.py:179]: @samsucik @dakshvar22 Does this explanation make sense?

Comment (samsucik) [reply rasa/core/featurizers/tracker_featurizers.py:179]: I think so.

**Comment by wochinge** [comment]
> (a) create a tracker_featurizers subdirectory with an __init__.py that imports the different tracker featurizers or

That seems to be the neatest option. If this is an overkill then I'd maybe skip the whole "moving to separate module" for now as I think there are probably more important things to do now.

**Comment by kedz** [comment]
> > (a) create a tracker_featurizers subdirectory with an **init**.py that imports the different tracker featurizers or
> 
> That seems to be the neatest option. If this is an overkill then I'd maybe skip the whole "moving to separate module" for now as I think there are probably more important things to do now.

@wochinge After discussing with @dakshvar22 and @JEM-Mosig, we are going to make splitting tracker_featurizers into separate modules [its own issue](https://github.com/RasaHQ/rasa/issues/9056) to be tackled later.

**Review by dakshvar22 — APPROVED** [review]
LGTM 🚀

**Review by wochinge — APPROVED** [review]
👍🏻

**Pull request merged**